components/kyma-environment-broker/internal/process/provisioning/azure_container_registry_step.go

```diff
@@ -0,0 +1,198 @@
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/common/hyperscaler"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/common/hyperscaler/azure"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/process"
+	processazure "github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/process/azure"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/runtime/components"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/uid"
+	"github.com/kyma-project/control-plane/components/provisioner/pkg/gqlschema"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	registryNamePrefix = "kyma"
+	registryAddress    = "azurecr.io"
+)
+
+var (
+	registryNameRegexp = regexp.MustCompile(`[^a-zA-Z0-9]`)
+)
+
+//go:generate mockery -name=UIDGenerator -output=automock -outpkg=automock -case=underscore
+type UIDGenerator interface {
+	Generate() string
+}
+
+// ensure the interface is implemented
+var _ Step = (*ProvisionAzureContainerRegistryStep)(nil)
+
+type ProvisionAzureContainerRegistryStep struct {
+	operationManager *process.ProvisionOperationManager
+	azure            processazure.ProviderContext
+	azureStepConfig  azure.StepConfig
+	uidSvc           UIDGenerator
+}
+
+func NewProvisionAzureContainerRegistryStep(os storage.Operations, hyperscalerProvider azure.HyperscalerProvider, accountProvider hyperscaler.AccountProvider, stepCfg azure.StepConfig, ctx context.Context) *ProvisionAzureContainerRegistryStep {
+	return &ProvisionAzureContainerRegistryStep{
+		operationManager: process.NewProvisionOperationManager(os),
+		azure: processazure.ProviderContext{
+			HyperscalerProvider: hyperscalerProvider,
+			AccountProvider:     accountProvider,
+			Context:             ctx,
+		},
+		azureStepConfig: stepCfg,
+		uidSvc:          uid.NewUIDService(),
+	}
+}
+
+func (s *ProvisionAzureContainerRegistryStep) Name() string {
+	return "Provision Azure Container Registry"
+}
+
+func (s *ProvisionAzureContainerRegistryStep) Run(operation internal.ProvisioningOperation, log logrus.FieldLogger) (internal.ProvisioningOperation, time.Duration, error) {
+	// check if step was finished successfully before, and the resource group name was persisted
+	if operation.Azure.ContainerRegistryCreated {
+		log.Info("Container Registry is already provisioned")
+		return operation, 0, nil
+	}
+
+	hypType := hyperscaler.Azure
```

Comment:
I was considering doing that but wasn't sure about it because most of the created objects (`pp`, `azureCfg`, `azureClient`) are used later on in the step, so it's not like they're only dependencies to create e.g. `azureClient`. Also, in case of error we would have to return the information about how the operation should be handled (retried? if so, how many times etc.). Lastly, the dependencies such as hyperscalerProvider would have to be passed to this boilerplate function. In the end, I didn't have enough time to think of a nice way to refactor this cleanly.

Reply:
> I was considering doing that but wasn't sure about it because most of the created objects (pp, azureCfg, azureClient) are used later on in the step, so it's not like they're only dependencies to create e.g. azureClient.

True, maybe we could achieve that using a shared struct. That way, the method could still access these created objects (via the struct in contrast to a method which returns a lot of objects).
But I leave it completely up to you if you think it makes sense. I was just thinking we could remove boilerplate code in the step as well as in the tests.

> Also, in case of error we would have to return the information about how the operation should be handled (retried? if so, how many times etc.). 

Mhh, can't we just check if the struct method returns an error and if yes return these 3 arguments ?